source/src/plugins/KolmogorovTest.cc

```diff
@@ -0,0 +1,171 @@
+	              "be used for TGraphs. While this test can take TH1s, the Kolmogorov test is intended for use on unbinned data, not histograms. "
+	              "See https://root.cern.ch/doc/master/classTH1.html#aeadcf087afe6ba203bcde124cfabbee4 for more information.";
+    }
+
+    //-------------------------------------------------------------------------------------------------
+
+    StatusCode KolmogorovTest::readSettings(const TiXmlHandle xmlHandle) {
+      RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::readParameter(xmlHandle, "UseUnderflow", m_useUnderflow));
+      RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::readParameter(xmlHandle, "UseOverflow", m_useOverflow));
+
+      return STATUS_CODE_SUCCESS;
+    }
+
+    std::string KolmogorovTest::getTestOptions(const bool isHistogram, bool useUnderflow, bool useOverflow) {
+      std::string optionsString = "";
+
+      if (useUnderflow and isHistogram) {
+	optionsString += "U";
+      }
+      if (useOverflow and isHistogram) {
+	optionsString += "O";
+      }
+      return optionsString;
+    }
+
+    void KolmogorovTest::userRun(MonitorElement* pMonitorElement, QualityTestReport &report) {
+      const bool hasObject = (pMonitorElement->object() != nullptr);
+      const bool hasReference = (pMonitorElement->reference() != nullptr);
+      const bool isObjHistogram = pMonitorElement->object()->InheritsFrom("TH1");
+      const bool isObjGraph = pMonitorElement->object()->InheritsFrom("TGraph");
+      const bool isRefHistogram = pMonitorElement->reference()->InheritsFrom("TH1");
+      const bool isRefGraph = pMonitorElement->reference()->InheritsFrom("TGraph");
+
+      if (!hasObject) {
+        report.m_message = "No object attached to monitor element";
+        report.m_quality = 0.f;
+        throw StatusCodeException(STATUS_CODE_INVALID_PTR);
+      }
+
+      if (!hasReference) {
+        report.m_message = "No reference attached to monitor element";
+        report.m_quality = 0.f;
+        throw StatusCodeException(STATUS_CODE_INVALID_PTR);
+      }
+
+      if ( (isObjGraph and !isRefGraph) or (isObjHistogram and !isRefHistogram) ) {
+        report.m_message = "The object and reference must be of the same type";
+        report.m_quality = 0.f;
+        throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
+      }
+
+      if (isObjGraph) {
+	TGraph* pGraph = pMonitorElement->objectTo<TGraph>();
+	TGraph* pReferenceGraph = pMonitorElement->referenceTo<TGraph>();
+	std::string m_options = KolmogorovTest::getTestOptions(isObjHistogram, m_useUnderflow, m_useOverflow);
+
+	int sizeGraph = pGraph->GetN();
+	int sizeRef = pReferenceGraph->GetN()  ;
+	double* pArrayGraph = pGraph->GetY();
+	double* pArrayRef = pReferenceGraph->GetY();
+
+	std::sort(&pArrayGraph[0], &pArrayGraph[sizeGraph]);
```

Comment:
ROOT documentation says : 
`Input: a,b: One-dimensional arrays of length na, nb, respectively. The elements of a and b must be given in ascending order. option is a character string to specify options "D" Put out a line of "Debug" printout "M" Return the Maximum Kolmogorov distance instead of prob`
I understand that the y values have to be sorted according to x ascending order. Am I wrong here ?

Reply:
It just says "in ascending order", so I assumed the numbers in the array had to be sorted, without reference to anything outside.

I also had difficulties with the sorting function, since `pArrayGraph` and `pArrayRef` are pointers to arrays, and the `std::sort` function works on an array, not a pointer to one. I just did some testing and I think it's actually sorting the _pointers_, in ascending order, not the objects!

Reply:
I still believe that the y values have to be sorted by x ascending order, meaning that given y0, y1, ..., yn, they have to be sorted with x0 > x1 > ... > xn. It would make no sense to compute the cumulative function with y-sorted values. In this case, the sorting algorithm is a bit more tricky. You would have use a `std::map` or a sorting algorithm that can handle this.